lib/ocpp/v201/monitoring_updater.cpp

```diff
@@ -0,0 +1,532 @@
+                << " to next timepoint: " << std::ctime(&dbg_time_aligned);
+
+    return aligned_timepoint;
+}
+
+EventData create_notify_event(int32_t unique_id, const std::string& reported_value, const Component& component,
+                              const Variable& variable, const VariableMonitoringMeta& monitor_meta) {
+    EventData notify_event;
+
+    notify_event.component = component;
+    notify_event.variable = variable;
+    notify_event.variableMonitoringId = monitor_meta.monitor.id;
+
+    notify_event.eventId = unique_id;
+    notify_event.timestamp = ocpp::DateTime();
+    notify_event.actualValue = reported_value;
+
+    if (monitor_meta.monitor.type == MonitorEnum::Periodic ||
+        monitor_meta.monitor.type == MonitorEnum::PeriodicClockAligned) {
+        notify_event.trigger = EventTriggerEnum::Periodic;
+    } else if (monitor_meta.monitor.type == MonitorEnum::Delta) {
+        notify_event.trigger = EventTriggerEnum::Delta;
+    } else if (monitor_meta.monitor.type == MonitorEnum::UpperThreshold ||
+               monitor_meta.monitor.type == MonitorEnum::LowerThreshold) {
+        notify_event.trigger = EventTriggerEnum::Alerting;
+    } else {
+        EVLOG_AND_THROW(std::runtime_error("Invalid monitor type"));
+    }
+
+    if (monitor_meta.type == VariableMonitorType::HardWiredMonitor) {
+        notify_event.eventNotificationType = EventNotificationEnum::HardWiredMonitor;
+    } else if (monitor_meta.type == VariableMonitorType::PreconfiguredMonitor) {
+        notify_event.eventNotificationType = EventNotificationEnum::PreconfiguredMonitor;
+    } else if (monitor_meta.type == VariableMonitorType::CustomMonitor) {
+        notify_event.eventNotificationType = EventNotificationEnum::CustomMonitor;
+    } else {
+        EVLOG_AND_THROW(std::runtime_error("Invalid monitor meta type"));
+    }
+
+    return notify_event;
+}
+
+MonitoringUpdater::MonitoringUpdater(std::shared_ptr<DeviceModel> device_model, notify_events notify_csms_events,
+                                     is_offline is_chargepoint_offline) :
+    device_model(std::move(device_model)),
+    notify_csms_events(std::move(notify_csms_events)),
+    is_chargepoint_offline(std::move(is_chargepoint_offline)),
+    monitors_timer([this]() { this->process_periodic_monitors(); }),
+    unique_id(0) {
+}
+
+MonitoringUpdater::~MonitoringUpdater() {
+    stop_monitoring();
+}
+
+void MonitoringUpdater::start_monitoring() {
+    // Bind function to this instance
+    auto fn = std::bind(&MonitoringUpdater::on_variable_changed, this, std::placeholders::_1, std::placeholders::_2,
+                        std::placeholders::_3, std::placeholders::_4, std::placeholders::_5, std::placeholders::_6,
+                        std::placeholders::_7);
+    device_model->register_variable_listener(std::move(fn));
+
+    monitors_timer.interval(std::chrono::seconds(1));
```

Comment:
What do you think about making this configurable? Looks like depending on the number of registered monitors and other configuration this monitor processing could be a quite resource intensive operation. Or is there some requirement in the spec for this?

Reply:
Understood, will provide this flexibility.

Reply:
Flexibility provided.